conans/test/functional/test_third_party_patch_flow.py

```diff
@@ -0,0 +1,199 @@
+                save("myfile.cpp", "mistake1\nsomething\nmistake2\nmistake3\nsome\n")
+                apply_conandata_patches(self)
+
+            def build(self):
+                for i in (1, 2, 3):
+                    if "mistake{}".format(i) in load(os.path.join(self.source_folder, "myfile.cpp")):
+                        raise Exception("MISTAKE{} BUILD!".format(i))
+        """)
+
+    client = TestClient()
+    client.save({"conanfile.py": conanfile,
+                 "conandata.yml": ""})
+    client.run("install .")
+    client.run("source .")
+    client.run("build .", assert_error=True)
+    assert "MISTAKE1 BUILD!" in client.out
+
+    # user decides to create patches, first init the repo
+    client.init_git_repo(folder="src")  # Using helper for user/email repo init
+    client.save({"src/myfile.cpp": "correct1\nsomething\nmistake2\nmistake3\nsome\n"})
+    # compute the patch
+    mkdir(os.path.join(client.current_folder, "patches"))
+    client.run_command("cd src && git diff > ../patches/patch1")
+    client.run_command("cd src && git add . && git commit -m patch1")
+    conandata = textwrap.dedent("""
+        patches:
+          "1.0":
+            - patch_file: "patches/patch1"
+        """)
+    client.save({"conandata.yml": conandata})
+
+    client.run("source .")
+    client.run("build .", assert_error=True)
+    assert "MISTAKE2 BUILD!" in client.out
+
+    client.save({"src/myfile.cpp": "correct1\nsomething\ncorrect2\nmistake3\nsome\n"})
+    # compute the patch
+    mkdir(os.path.join(client.current_folder, "patches"))
+    client.run_command("cd src && git diff > ../patches/patch2")
+    client.run_command("cd src && git add . && git commit -m patch1")
+
+    conandata = textwrap.dedent("""
+        patches:
+          "1.0":
+            - patch_file: "patches/patch1"
+            - patch_file: "patches/patch2"
+        """)
+    client.save({"conandata.yml": conandata})
+    client.run("source .")
+    client.run("build .", assert_error=True)
+    assert "MISTAKE3 BUILD!" in client.out
+
+    client.save({"src/myfile.cpp": "correct1\nsomething\ncorrect2\ncorrect3\nsome\n"})
+    # compute the patch
+    mkdir(os.path.join(client.current_folder, "patches"))
+    client.run_command("cd src && git diff > ../patches/patch3")
+    client.run_command("cd src && git add . && git commit -m patch1")
+
+    conandata = textwrap.dedent("""
+           patches:
+             "1.0":
+               - patch_file: "patches/patch1"
+               - patch_file: "patches/patch2"
+               - patch_file: "patches/patch3"
+           """)
+    client.save({"conandata.yml": conandata})
+    client.run("source .")
+    client.run("build .")
+    assert "conanfile.py (mypkg/1.0): Calling build()" in client.out
+
+    # of course create should work too
+    client.run("create .")
+    assert "mypkg/1.0: Created package" in client.out
+
+
+def test_third_party_overwrite_build_file():
```

Comment:
@lasote I have added these 2 more "flow" tests, to illustrate how this would behave for downloads and git clones. There is one workaround to do , but might be still worth it the approach.